Scrutins

njoyard opened the pull request:
TODO:
- [X] parse to return MP's and groups details
  - [X] Parse/clean demandeurs + extract group (Warning: demandeurs can be multiple => schema changed)
  - [X] fix groupes acronymes to match ND's (LAREM −> LREM)
  - [x] Fix votes mis au point missing groupe fields
- [x] Handle cases of intervention not found https://pad.regardscitoyens.org/p/nosdeputes-scrutins
  - [x] Scrutin 120 non trouvé dans les interventions de la séance 153 du 2017-09-25 16:00
  - [x] Scrutin 121 non trouvé dans les interventions de la séance 153 du 2017-09-25 16:00
  - [x] Scrutin 139 non trouvé dans les interventions de la séance 219 du 2017-10-10 15:00
  - [x] Scrutin 334 non trouvé dans les interventions de la séance 628 du 2017-12-06 15:00
  - [x] Scrutin 335 non trouvé dans les interventions de la séance 628 du 2017-12-06 15:00
  - [x] Scrutin 336 non trouvé dans les interventions de la séance 628 du 2017-12-06 15:00
  - [x] Scrutin 337 non trouvé dans les interventions de la séance 628 du 2017-12-06 15:00
  - [x] Scrutin 361 non trouvé dans les interventions de la séance 759 du 2018-01-23 15:00
  - [x] Scrutin 411 non trouvé dans les interventions de la séance 1013 du 2018-03-07 15:00
  - [x] Scrutin 438 non trouvé dans les interventions de la séance 1163 du 2018-04-03 09:30
  - [x] Scrutin 579 non trouvé dans les interventions de la séance 1374 du 2018-05-14 16:00
  - [x] Scrutin 739 non trouvé dans les interventions de la séance 1506 du 2018-05-31 21:30
  - [x] remonter à l'AN le problèmes des mauvaise séances référéncées https://github.com/regardscitoyens/Issues-Parlement/issues/44
  - [x] remonter à l'AN le problèmes des mauvais résultats dans les CR https://github.com/regardscitoyens/Issues-Parlement/issues/45
- [X] attach vote to specific debate
- [X] add presences to MPs voting without delegation
  - [x] check if non votant should be considered as presence or not (some cases just president, sometimes both presidents) if only president, no need as they always talk anyways
  - [x] validate non-votant not accounting for presences cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#issuecomment-413740999
  - [x] check presences added limited to actual votes and not rectif without previous position
  - [x] check presences added limited to date of first published delegations
  - [x] validate startdate for presence accounting cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115/commits/5a48a934aeae411ae654f8b9d5477f0696d27cfe
- [x] check missing scrutins (1106 sur 1118, motions notamment mais pas que)
  - [x] remonter à l'AN le problème https://github.com/regardscitoyens/Issues-Parlement/issues/40
  <s>- [ ] parse them manually from web ?...</s>
  - [x] plan case of motions with no vote contre, so unmatched by tagInterventions
- [x] check stats on mise au points, nonvotants, seance null, demandeur null, ...
  - [x] check delegations on scrutins non solennels
  - [x] remonter à l'AN le problème des seances manquantes https://github.com/regardscitoyens/Issues-Parlement/issues/42
  - [x] hardcorriger les seances manquantes ?
  - [x] remonter à l'AN le problème des demandeurs null https://github.com/regardscitoyens/Issues-Parlement/issues/43
  - [x] hardcorriger les demandeurs manquants ?
  - [x] investigate cases of position null (ok: all part of mises au point)
- [x] check data properly updated on load when scrutin modified (maybe not reset Seance and Tag Interv for instance since if it changed it can not be easily impacted on existing presences)
- [x] adjust logs for crons
- [x] check compare synthese data before/after
- [x] edit explications présences relevées in graphes
- [ ] add more doc/faq on votes (+ link in api to faq + link in activity to faq + in synthese instead of legend + link from activité présence hémicycle)

Extras:
- [ ] display scrutins in seances
- [ ] Parse/clean titles + extract scrutins object
  - [ ] attach vote to specific text
  - [ ] attach vote to specific amendment
  - [ ] tag sections with votes ?
  - [ ] retrotag TexteLoi adopted when identified ? cf #96
- [ ] build indicator of taux participation scrutins ? (distinction solennel/ordinaires?)
- [ ] build indicator of group fidelity ? (warning : non votants + mises au point)
- [ ] present list of an MP's known votes ?
- [ ] parse avis comm/gouv in debats ?
- [ ] add API points for votes ? cf #93 
- [ ] add presences from applauses ?
- [ ] viz votes ? (cf https://www.regardscitoyens.org/temp/sprint-scrutins/ )

TODO à la mise en prod : 
- [x] backup db
- [x] pull/merge
- [x] bin/updateDB8.sh
- [x] apt-get install python-html5lib python-requests
- [x] bin/load_scrutins
- [x] php symfony top:Deputes

njoyard committed:
Ajout script utilitaire de download open data AN

njoyard committed:
Mise en forme + dependances

njoyard committed:
Extraction referentiels open data + script presences scrutins

njoyard committed:
Schema scrutins

njoyard committed:
Maj schéma + parsing scrutins complet

njoyard committed:
Génération modele doctrine

njoyard committed:
Ajout task load scrutins

njoyard committed:
Ajout script update db

njoyard committed:
Recherche séance + tag interventions résultat scrutins

njoyard committed:
Détection délégations trop nombreuses

njoyard committed:
Typos

njoyard committed:
Ajout check nb scrutins/séance, recherche scrutins par votes et plus par ordre

njoyard committed:
Prise en compte date début délégations + typo

njoyard committed:
Ne plus utiliser les <table> pour trouver les scrutins

njoyard committed:
Accepter les virgules dans les scrutins

RouxRC committed:
Merge branch 'master' into scrutins

RouxRC commented on lib/model/doctrine/ParlementaireScrutin.class.php:
```diff
@@ -0,0 +1,29 @@
+
+    return $this->_set('parlementaire_id', $parl->id);
+  }
+
```
ptet renommer la fonction pour pas surcharger la fonction nativement générée

RouxRC commented on lib/model/doctrine/ParlementaireScrutin.class.php:
```diff
@@ -0,0 +1,29 @@
+
+  public function setScrutin($scrutin) {
+    return $this->_set('scrutin_id', $scrutin->id);
+  }
```
De même je soupçonne que c'est déjà le comportement natif sans cette fonction ;)

RouxRC commented on lib/model/doctrine/ParlementaireScrutin.class.php:
```diff
@@ -0,0 +1,29 @@
+      $this->parlementaire_groupe_acronyme,
+      'scrutin',
+      $this->Scrutin->getLinkSource(),
+      $this->position != 'nonVotant' && $this->mise_au_point_position == NULL
```
je pense que s'il y a bien une position passée sans délégation et une mise au point ça correspond bien à une présence, en revanche j'ai l'impression que par_delegation n'est jamais utilisé jusqu'ici pour confirmer les présences, je suppose donc que c'est une confusion ici et qu'il faut remplacer mise_au_point_position par par_delegation

RouxRC replied:
Par ailleurs non votant indique j'ai l'impression le président, il est donc a priori bien présent, je pense que le bon check est position not Null et par_delegation False

RouxRC commented on lib/model/doctrine/Scrutin.class.php:
```diff
@@ -0,0 +1,134 @@
+  public function setNumero($numero) {
+    return $this->_set('numero', $numero);
+  }
+
```
probablement inutile

RouxRC commented on lib/model/doctrine/Seance.class.php:
```diff
@@ -68,6 +68,35 @@ public function addPresenceLight($parlementaire_id, $groupe_acronyme, $type, $so
+        $presence->delete();
+      }
+    }
+
```
This raises an issue not adressed until now : we store the last source we find for each preuve of a same kind, hence if we have two votes for a same seance with no intervention for an MP, and if we then update one only of his votes, it will remove the presence forgetting the other vote not modified.
An easy turnout is to store one preuve by scrutin, but that means a different addPreuve function
A better one would be to change the global behaviours of Preuve to account all sources of a same kind of preuve, but that's a bit more work...

edit : if we consider "preuves" as documents, having one preuve by scrutin can make sense, so maybe the first solution makes sense in the end

RouxRC commented on lib/model/doctrine/Scrutin.class.php:
```diff
@@ -0,0 +1,134 @@
+         || !$parlscrutin->_set('par_delegation', $data->par_delegation)
+         || !$parlscrutin->_set('mise_au_point_position', $data->mise_au_point_position or NULL)) {
+          return FALSE;
+        }
```
Ces deux return FALSE au milieu du foreach sans Warning semblent un peu dangereux, en pratique il n'y a pas de raison qu'ils puissent arriver mais ptet raise plutot histoire d'avoir le suivi dans le catch plus bas

RouxRC committed:
merge recent dep fix

RouxRC committed:
cleanup, cf comments

RouxRC committed:
setup presences depending on position + no delegation

cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#pullrequestreview-146688706

RouxRC committed:
handle one preuve présence by source even for same type except for intervention

to avoid removing a presence corresponding to multiple proofs in the same seance (like 2 scrutins with only one rectified)
(and now keep first intervention instead of last as reference then)

cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#pullrequestreview-146694350

RouxRC commented:
longer todolist in #17

RouxRC committed:
lighter log for future crons

RouxRC committed:
include load_scrutins in daily loads

RouxRC committed:
adjust groupes names from opendata an to nd ones

RouxRC committed:
warn on missing scrutins in opendata AN

RouxRC committed:
opendata is updated only daily, no need to run load scrutins trice a day

RouxRC committed:
Adjust first date of delegations present

Il me semble qu'on avait discuté de ça au sprint mais je ne suis plus sûr.
On a deux dates de première apparition de délégations dans les données OpenData:
- en octobre 2017 pour deux scrutins solennels http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/183
- en mars 2018 pour une série de scrutins ordinaires http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/424 (ce n'est pas explicité sur la page web du scrutin mais dans les données Adrien Taquet est déclaré votant par délégation sur plusieurs autres scrutins ordinaires après celui là)

La question est : à partir de quand on considère un vote comme une présence s'il n'a pas le champ par_delegation à True.
J'ai l'impression qu'on peut considérer la première des deux dates comme la bonne. Mais le choix dans le code de la seconde me fait douter de ce qu'on s'était dit avec @njoyard. Peut-être est-ce aussi que les données auraient changé depuis et qu'ils ont fini par rétrorenseigner les données notamment pour faire leur application des sanctions, qui sait !? ;)
Du coup avis bienvenus ! cc @njoyard @teymour @tarnefys @Massiliane

PS: au final, on a dans l'OpenData : 14 scrutins solennels avec des délégations et... 22 scrutins ordinaires. Donc bien que ce soit exceptionnel, c'est bien toujours appliqué aussi sur ces scrutins contrairement à ce qu'on nous a toujours répondu à tort. ;)

RouxRC committed:
add checks on missing fields

RouxRC committed:
hardfix missing idJO on some Seances from OpenData AN

cf https://github.com/regardscitoyens/Issues-Parlement/issues/42

RouxRC committed:
first cleaning of demandeurs

RouxRC committed:
hardfix missing demandeurs

cf https://github.com/regardscitoyens/Issues-Parlement/issues/43

RouxRC committed:
fix warnings on missing fields

RouxRC committed:
change demandeur to plural in model + parse groupes

RouxRC committed:
add loaded dir to gitignore

RouxRC commented on lib/task/loadScrutinsTask.class.php:
```diff
@@ -0,0 +1,116 @@
+      if ($seances_manquantes > 0) {
+        echo "WARNING: $seances_manquantes séances non trouvées\n";
+      }
+    }
```
@njoyard le load de tous les scrutins marche d'un coup chez toi ?
en local chez moi ca crashe pour exces de ram régulièrement et il faut du coup le lancer en plein de fois, mais du coup les check finaux peuvent pas marcher :/

RouxRC replied:
En fait vu que de toute façon les fichiers qui petent restent à recharger les décomptes sont refaits à chaque fois donc c pas gênant

RouxRC committed:
catch potential error

RouxRC committed:
get back from opendata missing groupe of votant when only in mise au point

RouxRC committed:
get position_groupe from groupe for mises au point

RouxRC commented:
Quelques stats sur les données parsées : 

- **Résultats des scrutins :**
```
ls scrutins/*.json | while read f; do json $f | grep '"sort"'; done  | sort | uniq -c | sort -gr
    886     "sort": "rejeté",
    270     "sort": "adopté",
```

- **Types de scrutins :**
```
ls scrutins/*.json | while read f; do json $f | grep '"type"'; done  | sort | uniq -c | sort -gr
   1123     "type": "ordinaire"
     33     "type": "solennel"
```

- **Parts de scrutins avec délégation :**
```
ls scrutins/*.json | while read f; do json $f | grep "par_delegation"; done  | sort | uniq -c | sort -gr
 123130             "par_delegation": false,
   2659             "par_delegation": null,
   1769             "par_delegation": true,
```

- **Premiers scrutins avec délégation :**
```
solennel : 2017-10-24 http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/183
ordinaire : 2018-03-21 http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/424
```

- **Types de scrutins avec délégation :**
```
grep -c 'par_delegation": true' scrutins/*.json | grep -v ":0" | sed 's/:1//' | while read f; do json $f | grep '"type"'; done | sort | uniq -c | sort -gr
     22     "type": "ordinaire"
     14     "type": "solennel"
```

- **Usage des délégations par groupe :**
```
ls scrutins/*.json | while read f; do json $f | grep -B 3 'delegation": true' | grep groupe; done | sort | uniq -c | sort -gr
    949             "groupe": "LREM",
    334             "groupe": "LR",
    128             "groupe": "MODEM",
    102             "groupe": "NG",
    102             "groupe": "UAI",
     59             "groupe": "GDR",
     46             "groupe": "LFI",
     33             "groupe": "NI",
     16             "groupe": "LC",
```

- **Positions :**
```
ls scrutins/*.json | while read f; do json $f | grep '"position"'; done  | sort | uniq -c | sort -gr
  71314             "position": "contre",
  48145             "position": "pour",
   5764             "position": "abstention",
   2088             "position": "nonVotant",
    247             "position": null,
```

- **Mises au point :**
```
ls scrutins/*.json | while read f; do json $f | grep 'point_position'; done  | sort | uniq -c | sort -gr
 127086             "mise_au_point_position": null,
    257             "mise_au_point_position": "pour",
    167             "mise_au_point_position": "contre",
     31             "mise_au_point_position": "nonVotantVolontaire",
     13             "mise_au_point_position": "abstention",
      4             "mise_au_point_position": "nonVotant",
```

- **Positions avant mise au point :**
```
ls scrutins/*.json | while read f; do json $f | grep -A 4 'mise_au_point_position": "' | grep '"position"'; done | sort | uniq -c | sort -gr
    247             "position": null,
    117             "position": "contre",
     96             "position": "pour",
     10             "position": "abstention",
      2             "position": "nonVotant",
```
cas particuliers : 
```
count	position	délégation	mise_au_point

1987 	nonVotant 	null 		null
2 	nonVotant 	null 		contre

1 	pour 		false 		nonVotant
3 	contre 		false 		nonVotant

21 	contre 		false 		nonVotantVolontaire
10 	pour 		false 		nonVotantVolontaire

5 	pour 		false 		pour
```

- **Rares mises au point après délégation :**
```
ls scrutins/*.json | while read f; do json $f | grep -A 3 '"mise_au_point_position": "' | grep 'delegation'; done  | sort | uniq -c | sort -gr
    249             "par_delegation": null,
    219             "par_delegation": false,
      4             "par_delegation": true,

ls scrutins/*.json | while read f; do json $f | grep -A 3 '"mise_au_point_position": "' | grep -B 1 -A 1 'par_delegation": true' | grep position ; done 
  2        "mise_au_point_position": "abstention", "position": "pour",
  1        "mise_au_point_position": "pour", "position": "abstention",
  1        "mise_au_point_position": "contre", "position": "abstention",
```

- **Répartition positions/délégation/mises_au_point**:

```ls loaded/*.json | while read f; do json $f | grep -A 2 'point_position": '; done | tr "\n" " " | sed -r 's/(            "mise)/\n\1/g' | sed -r 's/^.+": ([^,]+),.+": ([^,]+),.+": ([^,]+),.*$/\3\t\2\t\1/' | grep . | sed 's/"//g' | sort | uniq -c ```

nombre | position | par_delegation | mise au point
-- | -- | -- | --
6 | abstention | false | contre
5436 | abstention | false | null
2 | abstention | false | pour
4 | abstention | null | null
1 | abstention | true | contre
131 | abstention | true | null
1 | abstention | true | pour
3 | contre | false | abstention
3 | contre | false | nonVotant
21 | contre | false | nonVotantVolontaire
69145 | contre | false | null
90 | contre | false | pour
50 | contre | null | null
416 | contre | true | null
2 | nonVotant | null | contre
1987 | nonVotant | null | null
2 | null | null | abstention
86 | null | null | contre
159 | null | null | pour
6 | pour | false | abstention
72 | pour | false | contre
1 | pour | false | nonVotant
10 | pour | false | nonVotantVolontaire
45379 | pour | false | null
5 | pour | false | pour
270 | pour | null | null
2 | pour | true | abstention
1209 | pour | true | null



- **Usage des mises au point par groupe :**
```
ls scrutins/*.json | while read f; do json $f | grep -B 1 'mise_au_point_position": "' | grep groupe; done | sort | uniq -c | sort -gr
    334             "groupe": "LREM",
     39             "groupe": "LR",
     33             "groupe": "MODEM",
     21             "groupe": "LFI",
     19             "groupe": "NG",
     10             "groupe": "GDR",
     10             "groupe": "LC",
      4             "groupe": "NI",
      2             "groupe": "UAI",
```

- **Demandeurs :**
```
ls scrutins/*.json | while read f; do python -m json.tool $f | grep '^        "' | grep -v ' "PA' | sed 's/\\u00e9/e/g' | sed 's/President[^\s]* du groupe [^A-Z]*\(.*\)/\L\1/' | sed 's/,$//'; done  | sort | uniq -c | sort -gr
    274         "les republicains"
    258         "la france insoumise"
    249         "gauche democrate et republicaine"
    237         "nouvelle gauche"
     72         "la republique en marche"
     59         "udi, agir et independants"
     29         "Conference des Presidents"
     21         "mouvement democrate et apparentes"
     18         "socialistes et apparentes"
     18         "les constructifs : republicains, udi, independants"
      1         "Gouvernement"
```

- **Groupes codemandeurs :**
```
SELECT count(id) as n, demandeurs_groupes_acronymes FROM scrutin WHERE  demandeurs_groupes_acronymes LIKE  '%|%' GROUP BY demandeurs_groupes_acronymes
n	demandeurs_groupes_acronymes
8	NG|GDR
6	LFI|GDR
4	NG|LFI
4	LR|GDR
4	GDR|LFI
3	NG|LR
3	LR|UAI
3	GDR|NG
2	LREM|LR
2	LREM|GDR
2	GDR|LR
1	UAI|LR
1	UAI|LFI
1	UAI|GDR
1	NG|LR|LREM|GDR
1	NG|LREM|GDR
1	NG|LFI|GDR
1	NG|LC
1	NG|GDR|LFI
1	LR|LREM
1	LR|LC|LREM
1	LREM|NG
1	LREM|MODEM
1	LREM|GDR|NG
1	LFI|NG|UAI
1	LFI|LR|NG
1	LFI|LR|GDR|NG
1	LFI|LREM
1	LFI|LC
1	LC|NG
1	LC|LFI
1	GDR|UAI

```

- **Répartition des groupes des votants :**
```
ls scrutins/*.json | while read f; do json $f | grep '"groupe"'; done  | sort | uniq -c | sort -gr
  78599             "groupe": "LREM",
  16790             "groupe": "LR",
   9767             "groupe": "MODEM",
   5377             "groupe": "NG",
   5328             "groupe": "LFI",
   3515             "groupe": "GDR",
   3431             "groupe": "NI",
   3038             "groupe": "UAI",
   1713             "groupe": "LC",
```

- **Séances avec le plus de scrutins :**
```
ls scrutins/*.json | while read f; do json $f | grep '"seance"'; done  | sort | uniq -c | sort -gr | head -n 10
     72     "seance": "20180263",
     27     "seance": "20180206",
     19     "seance": "20182008",
     18     "seance": "20180249",
     17     "seance": "20180248",
     15     "seance": "20171012",
     13     "seance": "20180256",
     13     "seance": "20180212",
     13     "seance": "20180203",
     12     "seance": "20180213",
```

RouxRC committed:
avoid saving a scrutin for which we can't find the intervention and reload only metadata + votes

RouxRC committed:
wrong link

RouxRC commented:
Les mises au point de type `nonVotantVolontaire` correspondent toutes à des votes exprimés sans délégation, pour lesquelles le relevé de vote indique dans les mises au point "M. XXX n’a pas pris part au scrutin" ([exemple](http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/262))
```
     21             "position": "contre",
     10             "position": "pour",
```
Il y a par ailleurs des cas identiques ou ces votes sont corrigés en `nonVotant` ([exemple](http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/470)), un autre avec ["Mme Paula Forteza a fait savoir qu’elle n'avait pas souhaité prendre part au scrutin"](http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/138) ou encore un cas où Caroline Fiat a dans l'opendata une mise au point en `nonVotant` qui n'apparait pas sur le [scrutin](http://www2.assemblee-nationale.fr/scrutins/detail/(legislature)/15/(num)/2).

Je trouve un peu bizarre de se déclarer a posteriori non votant, pourquoi pas juste abstention ? Du coup je ne sais pas trop s'il faut en déduire que le député n'était en fait pas présent et qu'il faut dans les cas de ces rectifs retirer sa présence ou pas...
des avis @njoyard @teymour @tarNeFyS @Massiliane ?

Se pose par ailleurs plus largement la question d'ajouter ou non une preuve de présence pour les positions "nonVotant" : il s'agit généralement du président de l'AN et des présidents de séance, leur présence est de toute façon déjà assurée en base, mais j'ai l'impression que de rugy est déclaré nonVotant sur tous les scrutins quasiment même lorsqu'il n'est pas en séance (je ne crois pas qu'il assiste à des séances autrement qu'en les présidant). Du coup après avoir hésité j'aurais finalement tendance à ne pas compter les nonVotants comme preuve de présence. Avis bienvenus également !

Les auteurs de toutes les positions `nonVotants` sont répartis comme suit :
```
ls scrutins/*.json | while read f; do json $f | grep -B 4 '"position": "nonVotant"' | grep 'PA'; done  | sort | uniq -c | sort -gr
   1106         "PA332747": {    # De Rugy, tous les votes
    155         "PA720746": {    # les 6 vice-prez actuels
    146         "PA721824": {
    137         "PA720622": {
    123         "PA1874": {
    102         "PA605991": {
    101         "PA267801": {
     61         "PA722150": {   # les 2 vice prez LREM en rab du début
     41         "PA719850": {
      9         "PA719890": {   # une vice prez temporaire
      1         "PA721872": {   # que des ministres qui ont du s'abstenir sur un vote lors de leur période d'un mois après la nomination au gouv
      1         "PA721560": {
      1         "PA719914": {
      1         "PA717167": {
      1         "PA607395": {
      1         "PA605131": {
      1         "PA337633": {
      1         "PA331481": {
```

RouxRC committed:
hardfix errors association seance from opendata AN

cf https://github.com/regardscitoyens/Issues-Parlement/issues/44

RouxRC committed:
fix more cases of bad seances listed in opendata

RouxRC committed:
match latest table in an intervention

notably to handle manual correction cases such as
https://www.nosdeputes.fr/15/seance/219#table_486

RouxRC committed:
oups wrong interpretation of preg_match_all results architecture

RouxRC committed:
oups debug

RouxRC committed:
no need to search for scrutins in committees

RouxRC commented:
Quelques stats sur les présences en base avant/après le load des scrutins :

```
1106 scrutins (sur 1118)
124 499 votes
89 304 générant une preuve de présence
22 090 nouvelles présences identifiées
environ 40 par député

avant/après
63 143/85 233 présences
112 076/201 380 preuves
```

- Répartition des semaines d'activité avant/après ([données](https://www.regardscitoyens.org/temp/semaines-activité-avant-apres-votes.ods)) :

![https://www.regardscitoyens.org/temp/semaines-activité-avant-apres-votes.png](https://www.regardscitoyens.org/temp/semaines-activité-avant-apres-votes.png)

154 au dessus de 37 avant / 217 au dessus de 38 après
151 au dessous de 30 avant / 175 au dessous de 34 après

value | avant | après
-- | -- | --
43 | 2 | 2
42 | 4 | 5
41 | 13 | 27
40 | 22 | 38
39 | 30 | 66
38 | 47 | 79
37 | 46 | 68
36 | 52 | 57
35 | 55 | 53
34 | 54 | 36
33 | 31 | 28
32 | 35 | 20
31 | 28 | 19
30 | 27 | 18
29 | 26 | 10
28 | 14 | 5
27 | 16 | 6
26 | 10 | 6
25 | 7 | 6
24 | 8 | 2
23 | 11 | 4
22 | 3 | 5
21 | 6 | 2
20 | 3 | 1
19 | 3 | 4
18 | 4 | 1
17 | 3 | 0
16 | 3 | 0
15 | 1 | 0
14 | 2 | 1
13 | 1 | 0
12 | 0 | 0
11 | 1 | 0
10 | 0 | 0
9 | 1 | 0
8 | 0 | 1
7 | 0 | 0
6 | 0 | 0
5 | 0 | 0
4 | 0 | 0
3 | 1 | 0

RouxRC committed:
don't account votes 'NonVotant' as présents

cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#issuecomment-413740999

RouxRC committed:
update legends of semaines d'activité to signify scrutins are accounted

RouxRC committed:
Merge branch 'master' into scrutins

RouxRC committed:
refacto indicateurs titles and descr for shared use between synthese and MPs pages

Massiliane commented:
Merci @RouxRC  et @njoyard pour votre travail sur les scrutins ! 
Je vais essayer, dans la limite de mes capacités de vous donner mon avis.

- Sur ce point des non-votants : 
"Il y a par ailleurs des cas identiques ou ces votes sont corrigés en nonVotant
Je trouve un peu bizarre de se déclarer a posteriori non votant, pourquoi pas juste abstention ?"
Pour ma part, je pense qu'il ne faut pas les prendre en compte comme des présences. En fait, c'est souvent le cas du député qui ne fait pas de délégation de vote, qui ne veut pas participer au vote et pour lequel le groupe a déjà prévu une délégation sans son accord. Du coup, le fait de venir faire rectifier pour moi montre que le député ne voulait pas être "comptabilisé" comme ayant participé à cette activité. 
Pour moi, dans les cas où le député ne participe pas à l'activité (en votant) on devrait respecter sa volonté de ne pas participer.

- Pour ton deuxième point sur les preuves de présence des non votants : 
"Se pose par ailleurs plus largement la question d'ajouter ou non une preuve de présence pour les positions "nonVotant""
Je n'ai pas trop d'avis mais si effectivement leur présence est déjà détecté lorsqu'ils président la séance, je ne vois pas pourquoi on devrait prendre en compte le "nonVotant" dans ce cadre là.

Merci encore !

DavidOpVa commented:
Bravo pour le travail.

Ok pour moi de dire que les non-votant ne sont pas comptabilité comme présent au vu des raisons invoquées en général. Je pense par ailleurs qu'il est important de conserver un fonctionnement simple et facilement intuitable, et du coup je préfère éviter des solutions trop complexes où l'on compterait des présences seulement pour certain type de non-votant.

boogheta commented:
Verdicts a priori de la réunion garbage deu 03/09/18 : 

NonVotants :
- on demande à l'AN quelle est la différence entre abstention, nonVotant et nonVotantVolontaire pour les rectifications
- position : nonVotant ne compte jamais comme présents
- mise_au_point : nonVotant ne compte pas comme présence
- mise_au_point : nonVotantVolontaire et abstention comptent comme présence si pas de délégation (sous réserve de validation par l'AN qu'il s'agit de l'application de l'alinéa 2, 7° article 13 de l'instruction du bureau http://www.assemblee-nationale.fr/connaissance/instruction.asp#igb_13 )

Date de début de prise en compte des délégations : 
- on intégre des présences pour les solennels sans délégation à partir d'octobre 2017
- on intégre des présences pour tous sans délégation à partir de mars 2018

RouxRC committed:
Merge branch 'master' into scrutins

Conflicts:
	.gitignore
	ansible/roles/cpc.install/templates/web_Dockerfile.j2

RouxRC committed:
don't integrate scrutins with seances missing

RouxRC commented:
![delegations-ordinaires](https://user-images.githubusercontent.com/1153255/45601546-1090c900-ba0f-11e8-86f6-2efdb2cf8de7.png)
![delegations-solennels](https://user-images.githubusercontent.com/1153255/45601547-1090c900-ba0f-11e8-81f5-60a957416628.png)

Suite à la dernière discussion, j'ai tracé l'indication de délégations en séparant scrutins solennels et ordinaires : 
- sur les ordinaires c'est assez épisodique et il est difficille d'en tirer autre chose que la date minimale de mars 2018
- sur les solennels c'est plus intéressant : à partir de février 2018 c'est systématique et semble une bonne date minimale, avant cette date des délégations ne sont indiquées que pour 3 cas particuliers en octobre et décembre 2017 avec pourtant plusieurs solennels au meme ordre de grandeur de votants entre octobre et décembre

Du coup je propose d'amender la décision de prendre compte des présences comme suit : 
- on intégre des présences pour tous les scrutins publics et solennels à partir du 13 février 2018 
- on intégre des présences pour les 3 scrutins solennels particuliers indiquant des délégations <=> on intégre des présences pour tous les scrutins ayant au moins une délégation

RouxRC committed:
Merge branch 'master' into scrutins

RouxRC committed:
add warnings on position identical to mise_au_point

RouxRC committed:
adjust period of accounting presences via scrutins depending on delegations

cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#issuecomment-421844588

RouxRC committed:
adjust accounting presences for special cases of mise_au_point

cf https://github.com/regardscitoyens/nosdeputes.fr/pull/115#issuecomment-418172211

RouxRC committed:
woups

RouxRC committed:
set the field for all to avoid php warnings

RouxRC committed:
[WIP] Edit FAQ

RouxRC commented:
Réponse de la Division de l’Information multimédia au Service de la Communication et de l’Information multimédia de l'Assemblée nationale : 

- concernant la publicité des délégations : 
<blockquote>Les analyses de scrutins ne font apparaitre les délégations qu’ à compter du 20 mars 2018. La  fiabilité des données sur les délégations ne peut donc être assurée  qu’à compter de cette date. S’agissant des scrutins antérieurs au 20 mars 2018, la mention des délégations résultait d’un « effet de bord » de re -publications de l’analyse politique de certains scrutins ; les  scrutins antérieurs à cette date ont été à nouveau publiés sans mention des délégations.</blockquote>

Il nous faut donc : 
- [x] changer dans le code la date minimale de prise en compte des présences
- [x] supprimer en base en prod les preuves de présence et présence en trop ajoutées pour les scrutins publics sans délégation avant le 19 mars 2018

- concernant les mises au point : 
<blockquote><ul>
<li>Abstention : l’abstention est une position de vote, cette mise au point est utilisée dans le cas où le député s’est trompé dans sa position de vote ;</li>
<li>Non votant volontaire : Cette mise au point correspond à la situation d’un député qui, présent lors d’un scrutin, ne souhaitait pas prendre part au vote (application du 7° de l’article 13 de l’instruction générale du bureau) ;</li>
<li>Non votant : cette mise au point correspond à une erreur de manipulation du plot d’un député, absent lors du scrutin.</li>
</ul></blockquote>

Rien à changer donc de ce côté.

RouxRC commented:
Cela concerne 30 scrutins : 
- 2 solennels pour lesquels les délégations sont renseignées donc pour lesquelles les présences associées peuvent être conservées
- 28 ordinaires pour lesquels il faut supprimer les présences associées : entre 390 et 406 et entre 408 et 418 

```
ls {loaded,scrutins}/*.json | while read f; do json $f | grep "date\|nb_deleg\|numero"; done | tr  "\n" " " | sed 's/"date"/\n"date"/g'
"date": "2018-02-13",     "nb_delegations": 107,     "numero": 389,     
"date": "2018-02-13",     "nb_delegations": 0,     "numero": 390,     
"date": "2018-02-13",     "nb_delegations": 0,     "numero": 391,     
"date": "2018-02-13",     "nb_delegations": 0,     "numero": 392,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 393,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 394,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 395,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 396,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 397,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 398,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 399,     
"date": "2018-02-14",     "nb_delegations": 0,     "numero": 400,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 401,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 402,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 403,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 404,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 405,     
"date": "2018-02-15",     "nb_delegations": 0,     "numero": 406,     
"date": "2018-02-20",     "nb_delegations": 109,     "numero": 407,     
"date": "2018-03-06",     "nb_delegations": 0,     "numero": 408,     
"date": "2018-03-06",     "nb_delegations": 0,     "numero": 409,     
"date": "2018-03-06",     "nb_delegations": 0,     "numero": 410,     
"date": "2018-03-07",     "nb_delegations": 0,     "numero": 411,     
"date": "2018-03-07",     "nb_delegations": 0,     "numero": 412,     
"date": "2018-03-07",     "nb_delegations": 0,     "numero": 413,     
"date": "2018-03-08",     "nb_delegations": 0,     "numero": 414,     
"date": "2018-03-08",     "nb_delegations": 0,     "numero": 415,     
"date": "2018-03-08",     "nb_delegations": 0,     "numero": 416,     
"date": "2018-03-08",     "nb_delegations": 0,     "numero": 417,     
"date": "2018-03-08",     "nb_delegations": 0,     "numero": 418,     
```


- Avant / Après :
présences 82 714 / 82 283 => 431 en moins
preuves 192 852 / 192,025 => 827 en moins
